mars/tensor/base/searchsorted.py

```diff
@@ -0,0 +1,341 @@
+            in_chunks = [c]
+            if input_len == 2:
+                in_chunks.append(v.chunks[0])
+            chunks.append(chunk_op.new_chunk(in_chunks, shape=v_shape,
+                                             index=c.index, order=out.order))
+
+        while len(chunks) > combine_size:
+            new_chunks = []
+            it = itertools.count(0)
+            while True:
+                j = next(it)
+                to_combine = chunks[j * combine_size: (j + 1) * combine_size]
+                if len(to_combine) == 0:
+                    break
+
+                new_chunks.append(
+                    cls._combine_chunks(to_combine, op, _STAGES[1], v, j))
+            chunks = new_chunks
+
+        chunks = [cls._combine_chunks(chunks, op, _STAGES[2], v, 0)]
+
+        new_op = op.copy().reset_key()
+        return new_op.new_tensors(op.inputs, out.shape,
+                                  chunks=chunks, nsplits=((s,) for s in out.shape))
+
+    @classmethod
+    def tile(cls, op):
+        a = op.inputs[0]
+        out = op.outputs[0]
+        input_len = len(op.inputs)
+        if input_len == 1:
+            v = op.values
+        else:
+            v = op.inputs[1]
+            v = v.reshape(v.shape).single_tiles()
+
+        if len(a.chunks) == 1:
+            return cls._tile_one_chunk(op, a, v, out)
+        return cls._tile_tree_reduction(op, a, v, out)
+
+    @classmethod
+    def _execute_without_stage(cls, xp, a, v, op):
+        return xp.searchsorted(a, v, side=op.side)
+
+    @classmethod
+    def _execute_map(cls, xp, a, v, op):
+        indices = xp.atleast_1d(xp.searchsorted(a, v, side=op.side))
+        data_indices = indices.copy()
+        data_indices = xp.subtract(data_indices, 1, out=data_indices,
+                                   where=data_indices >= len(a))
+        data = a[data_indices]
+        if op.offset > 0:
+            indices = xp.add(indices, op.offset, out=indices)
+        if np.isscalar(v):
+            indices, data = indices[0], data[0]
+
+        return indices, data
+
+    @classmethod
+    def _execute_combine(cls, xp, a, v, op):
+        inp_indices, inp_data = a
+        if np.isscalar(v):
+            ind = xp.searchsorted(inp_data, v, side=op.side)
+            if ind >= len(inp_data):
+                ind -= 1
+            return inp_indices[ind], inp_data[ind]
+        else:
+            ret_indices = np.empty(v.shape, dtype=int)
+            ret_data = np.empty(v.shape, dtype=inp_data.dtype)
+            for idx in itertools.product(*(range(s) for s in v.shape)):
+                ind = xp.searchsorted(inp_data[(slice(None),) + idx], v[idx], side=op.side)
+                if ind >= len(inp_indices):
+                    ind -= 1
+                ret_indices[idx] = inp_indices[(ind,) + idx]
+                ret_data[idx] = inp_data[(ind,) + idx]
+            return ret_indices, ret_data
+
+    @classmethod
+    def _execute_reduce(cls, xp, a, v, op):
```

Comment:
This method seems almost same as `_execute_combine`, should they share the same code ?

Reply:
It's hard to extract common part.